test/bin/gen_dune_rules.ml

```diff
@@ -0,0 +1,151 @@
+(*
+ * Copyright (c) 2019 Craig Ferguson <me@craigfe.io>
+ *
+ * Permission to use, copy, modify, and distribute this software for any
+ * purpose with or without fee is hereby granted, provided that the above
+ * copyright notice and this permission notice appear in all copies.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *)
+
+let is_testcase filename =
+  let valid_extensions = [".md"; ".t"] in
+  List.exists (Filename.check_suffix filename) valid_extensions
+
+let read_file filename =
+  let lines = ref [] in
+  let chan = open_in filename in
+  try
+    while true do
+      lines := input_line chan :: !lines
+    done;
+    !lines
+  with End_of_file ->
+    close_in chan;
+    List.rev !lines
+
+let expected_file file =
+  let e = file ^ ".expected" in
+  if not (Sys.file_exists e) then
+    Fmt.failwith "No %s file found when generating an expect rule" e;
+  e
+
+(* Copied from Filename (stdlib) for pre-4.04 compatibility *)
+let chop_extension name =
+  let is_dir_sep s i = match Sys.os_type with
+    | "Unix" -> s.[i] = '/'
+    | "Win32" | "Cygwin" ->
+      let c = s.[i] in
+      c = '/' || c = '\\' || c = ':'
+    | _ -> assert false
+  in
+  let rec search_dot i =
+    if i < 0 || is_dir_sep name i then invalid_arg "Filename.chop_extension"
+    else if name.[i] = '.' then String.sub name 0 i
+    else search_dot (i - 1) in
+  search_dot (String.length name - 1)
+
+let options_of_file file =
+  let options_file = chop_extension file ^ ".opts" in
+  if not (Sys.file_exists options_file) then
+    [""]
+  else
+    read_file options_file
+
+let pp_options =
+  Fmt.(list ~sep:(const string " ") string)
+
+(** Tests that the output of 'ocaml-mdx test [options] <file>' is equal to the
+    contents of '<file>'. Uses the rule generator build into ocaml-mdx. *)
+let test_fixpoint_rule file =
+  let options = options_of_file file in
+  let command =
+    let args = file::options in
+    Fmt.strf "ocaml-mdx rule %a" pp_options args
```

Comment:
This is calling `ocaml-mdx` I have on my system if I `dune clean`. You should add `(package mdx)` on every rules calling this program.

Reply:
Thanks :slightly_smiling_face: It's only necessary in the `test_fixpoint/dune` case, since other rule generation patterns don't rely on `ocaml-mdx rule`. Have added it there.

Reply:
I'm not super convinced by this one. I feel like we should either test `ocaml-mdx test` or `ocaml-mdx rule` but not both at the same time.

If I want to test `ocaml-mdx rule`, I will compare its output with a `.expected` file. 

None of the tests in `test_fixpoint` actually require calling `ocaml-mdx rule` and this could just be another case of `test_expect` from what I can see. I might be wrong so let me know if it is in fact required but if it's not, I think we could detect whether the `.expected` file exists and compare to input on if it doesn't!

Reply:
Another advantage of my suggested approach is that we'd get rid of the annoying `execvp` errors!

Reply:
Regarding testing ocaml-mdx rule separately: I agree completely, but think this is something we should do _next_. The test directories are structured so that a `test/rule_expect` will fit in there nicely. The current state of affairs is how it was being done before, and I kept it this way to keep the diff smaller (not introducing a bunch of additional tests in the middle of a refactor).

Regarding making `test_fixpoint` a conditional behaviour of `test_expect` in the case where the expected file doesn't exist: I'd rather not do this. I'd rather the behaviour was kept explicit, and that each `test_foo` be kept as simple as possible. The code can be simplified in another way by testing `ocaml-mdx rule` generation separately and then not depending on `ocaml-mdx rule` to generate tests of other properties of the binary.

Reply:
I agree about keeping the behaviour explicit!

My main point was that I think you could just make the current `test_fixpoint_rule` very similar to `test_expect_rule`. I agree it changes how things were done but I don't think it will change what's tested too much and we could add proper `ocaml-mdx rule` tests in a separate PR.

Let's merge it like this, I'll improve it later on, you already had to rebase this quite a few times by now!